cpp/community_detection_module/algorithm/label_propagation.cpp

```diff
@@ -0,0 +1,291 @@
+#include <cmath>
+#include <mg_graph.hpp>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "label_propagation.hpp"
+
+namespace LabelRankT {
+
+///@brief Checks if vector is subset of another vector.
+///
+///@tparam T -- set element data type
+///
+///@param v1 -- subset candidate
+///@param v2 -- vector against which the check is performed
+///
+///@return -- whether v1 is subset of v2
+template <class T>
+bool is_subset(std::vector<T> const& v1, std::vector<T> const& v2) {
+  for (const auto& element : v1) {
+    if (std::find(v2.begin(), v2.end(), element) == v2.end())
+      return false;
+  }
+
+  return true;
+}
+
+double LabelRankT::get_weight(std::uint64_t node_i_id,
+                              std::uint64_t node_j_id) {
+  // WORK IN PROGRESS
+
+  return DEFAULT_WEIGHT;
+}
+
+void LabelRankT::set_structures(std::uint64_t node_id) {
+  times_updated[node_id] = 0;
+  std::unordered_map<std::uint64_t, double> node_label_Ps;
+
+  double sum_w_i = w_selfloop;
+  for (const auto node_j : graph->Neighbours(node_id))
+    sum_w_i += get_weight(node_j.node_id, node_id);
+
+  // add self-loop
+  node_label_Ps[node_id] = w_selfloop / sum_w_i;
+
+  // add other edges
+  for (const auto node_j : graph->Neighbours(node_id))
+    node_label_Ps[node_j.node_id] =
+        get_weight(node_id, node_j.node_id) / sum_w_i;
+
+  label_Ps[node_id] = node_label_Ps;
+  sum_w[node_id] = sum_w_i;
+
+  return;
+}
+
+void LabelRankT::remove_deleted_nodes(
+    std::unordered_set<std::uint64_t>& deleted_nodes_ids) {
+  for (const auto node_id : deleted_nodes_ids) {
+    label_Ps.erase(node_id);
+    sum_w.erase(node_id);
+    times_updated.erase(node_id);
+  }
+
+  return;
+}
+
+void LabelRankT::reset_times_updated() {
+  for (auto [node_id, _] : times_updated) times_updated[node_id] = 0;
+  return;
+}
+
+std::uint64_t LabelRankT::get_label(std::uint64_t node_id) {
+  std::uint64_t node_label;
+  double max_P = 0;
+
+  for (const auto [label, P] : label_Ps[node_id]) {
+    if (P > max_P) {
+      max_P = P;
+      node_label = label;
+    }
+  }
+
+  return node_label;
+}
+
+std::vector<std::uint64_t> LabelRankT::most_probable_labels(
+    std::uint64_t node_id) {
+  double max_P = 0;
+
+  for (const auto [label, P] : label_Ps[node_id]) {
+    if (P > max_P) max_P = P;
+  }
+
+  std::vector<std::uint64_t> most_probable_labels = {};
+
+  for (const auto [label, P] : label_Ps[node_id]) {
+    if (P == max_P) most_probable_labels.push_back(label);
+  }
+
+  return most_probable_labels;
+}
```

Comment:
Maybe ordered map where we check only maximum values?

Reply:
As `propagate()` is the bottleneck of the algorithm and heavily retrieves label probabilities, I ultimately made the design choice to use an unordered map because of it having O(1) retrieval.

Reply:
But you  are always having 2 iterations over whole set. In the other case, you would only have to iterate through first few similar elements.

Reply:
According to the paper (https://arxiv.org/pdf/1305.2006.pdf), the set that is iterated upon is very small thanks to the cutoff operation:
> More importantly, Φr is shown empirically to
reduce the space complexity efficiently, from quadratic to
linear. The average number of labels in each node is typically less than 3.0 for r = 0.1.

Does this check out or is an ordered map still preferable?

Reply:
If less than 3, than totally fine :smile: